subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,619 @@
+ * directly without intermediate copying.
+ *
+ * Limitations:
+ *  - Remote wakeup before IN transfer is not yet supported.
+ *  - H4 transport layer is not yet supported
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/usb_ch9.h>
+#include <zephyr/drivers/usb/udc.h>
+
+#include <zephyr/net/buf.h>
+
+#include <zephyr/bluetooth/buf.h>
+#include <zephyr/bluetooth/hci_raw.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/hci_vs.h>
+#include <zephyr/drivers/bluetooth/hci_driver.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hci, CONFIG_USBD_BT_HCI_LOG_LEVEL);
+
+#define BT_HCI_SUBCLASS		0x01
+#define BT_HCI_PROTOCOL		0x01
+
+/*
+ * The actual endpoint addresses may differ from the following because
+ * the resources may be reallocated by the stack depending on the number
+ * of functions in a configuration and the properties of the controller.
+ */
+#define BT_HCI_EP_EVENTS		0x81
+#define BT_HCI_EP_ACL_DATA_IN		0x82
+#define BT_HCI_EP_ACL_DATA_OUT		0x02
+#define BT_HCI_EP_VOICE_IN		0x83
+#define BT_HCI_EP_VOICE_OUT		0x03
+
+#define BT_HCI_EP_MPS_EVENTS		16
+#define BT_HCI_EP_MPS_ACL_DATA		0	/* Get maximum supported */
+#define BT_HCI_EP_MPS_VOICE		9
+
+#define BT_HCI_EP_INTERVAL_EVENTS	1
+#define BT_HCI_EP_INTERVAL_VOICE	3
+
+#define BT_HCI_CLASS_ENABLED		0
+#define BT_HCI_ACL_RX_ENGAGED		1
+
+static K_FIFO_DEFINE(bt_hci_rx_queue);
+static K_FIFO_DEFINE(bt_hci_tx_queue);
+
+/*
+ * Transfers through three endpoints proceed in a synchronous manner,
+ * with maximum packet size of high speed bulk endpoint.
+ *
+ * REVISE: global (bulk, interrupt, iso) specific pools would be more
+ * RAM usage efficient.
+ */
+NET_BUF_POOL_FIXED_DEFINE(bt_hci_ep_pool,
+			  3, 512,
+			  sizeof(struct udc_buf_info), NULL);
+/* HCI RX/TX threads */
+static K_KERNEL_STACK_DEFINE(rx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+static K_KERNEL_STACK_DEFINE(tx_thread_stack, CONFIG_USBD_BT_HCI_TX_STACK_SIZE);
+static struct k_thread tx_thread_data;
+
+struct bt_hci_data {
+	struct net_buf *acl_buf;
+	uint16_t acl_len;
+	struct k_sem sync_sem;
+	atomic_t state;
+};
+
+/*
+ * Make supported device request visible for the stack
+ * bRequest 0x00 and bRequest 0xE0.
+ */
+static const struct usbd_cctx_vendor_req bt_hci_vregs =
+	USBD_VENDOR_REQ(0x00, 0xe0);
+
+/*
+ * We do not support voice channels and we do not implement
+ * isochronous endpoints handling, these are only available to match
+ * the recomendet configuration in the Bluetooth specification and
+ * avoid issues with Linux kernel btusb driver.
+ */
+struct usbd_bt_hci_desc {
+	struct usb_association_descriptor iad;
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_in_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_if_descriptor if1_0;
```

Comment:
Bluetooth specification requires 4 alternate settings with varying max packet size. Here you have only 2.

Reply:
It used to require in v2.1 and v3.0, but now it only recommends 4 alternate settings. Older versions can also just recommend it when they implement Core Spec Addendum 2.